### PR TITLE
Rework CI tests to allow for ignoring, provider, and more

### LINKED ISSUE #528: Allow for adhoc testing of different targets/drivers
Currently we have the following targets we test against:
centos6, centos7, fedora26 and fedora27

Of those targets we then test the following drivers:
dummy, duffy, libvirt, aws and beaker

What this issue is about is to allow for finer control over which tests and drivers are executed for which target.

One approach would be to create a test.d directory that would contain an ordered list of tests to execute.  for example:
01_up
02_check_inventory
03_destroy

The default would be to run all these tests on every target/driver combination.  But allow driver/tests to be blacklisted to not run if not supported.
centos6/aws/tests_to_ignore.manifest:
// ignore all tests, effectively disable aws testing on centos6
.*
fedora26/duffy/tests_to_ignore.manifest:
// ignore check_inventory test and use regex so it doesn't matter what order the test is
[0-9]+_check_inventory

The updated harness should also support having a list of drivers/tests passed in as env variables so that the upper level CI infrastructure can skip tests that do not have any code changes made to them.

This is just a starting proposal and could change.

### LINKED ISSUE #576: Template data doesn't get rendered in topology file
linchpin version 1.5.3

I have following topology file where I use list of variables: image, executor_name, network, etc
```yaml
# linchpin-ws/topologies/my-cluster.yml 
---
topology_name: my_cluster
resource_groups:
  - resource_group_name: openstack
    resource_group_type: openstack
    resource_definitions:
      - name: "{{ executor_name }}-master"
        type: os_server
        flavor: w1.large
        image: "{{ image }}"
        count: 1
        keypair: "{{ keypair }}"
        networks:
          - "{{ network }}"
        fip_pool: "{{ fip_pool }}"
        security_groups: "{{ security_group }}"
      - name: "{{ executor_name }}-node"
        type: os_server
        flavor: w1.large
        image: "{{ image }}"
        count: 2
        keypair: "{{ keypair }}"
        networks:
          - "{{ network }}"
        fip_pool: "{{ fip_pool }}"
        security_groups: "{{ security_group }}"
    credentials:
      filename: clouds.yml
      profile: "{{ os_cloud }}"
```

I want to pass these variables from template-data file as following
```yaml
# linchpin-ws/Data.my_cluster.yml 
---
executor_name: my-executor-lbednar
keypair: my-executor-lbednar-net-key
network: my-executor-lbednar-net
security_group: my-executor-lbednar-secgroup
image: rhel-7.4-server-x86_64-latest
fip_pool: 10.8.240.0
os_cloud: centralci
```
Linchpin fails to find `image` parameter, because `image: "{{ image }}` in topology file was rendered into `image: ""` so it looks like that template-data are not used to render topology file.
```
$ linchpin -vvv -w linchpin-ws/ --creds-path /home/my-qe-jenkins/.config/openstack -d @Data.my_cluster.yml up my_cluster
TASK [openstack : provision/teardown os_server resources with provided auth] ***
task path: /home/my-qe-jenkins/jenkins/workspace/deploy-my-cluster-lbednar/.venv/lib/python2.7/site-packages/linchpin/provision/roles/openstack/tasks/provision_os_server.yml:2
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "api_timeout": 99999, 
            "auth": {
                "changed": true, 
                "failed": false, 
                "gid": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "group": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "mode": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "output": {
                    "clouds": {
                        "centralci": {
                            "auth": {
                                "auth_url": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                                "project_name": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                                "username": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
                            }
                        }
                    }
                }, 
                "owner": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "params": {
                    "cred_path": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                    "cred_type": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                    "driver": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                    "filename": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
                }, 
                "path": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "secontext": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "size": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "state": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "uid": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
            }, 
            "auth_type": null, 
            "auto_ip": true, 
            "availability_zone": null, 
            "boot_from_volume": false, 
            "boot_volume": null, 
            "cacert": null, 
            "cert": null, 
            "cloud": null, 
            "config_drive": false, 
            "count": 1, 
            "delete_fip": false, 
            "endpoint_type": "public", 
            "flavor": "w1.large", 
            "flavor_include": null, 
            "flavor_ram": null, 
            "floating_ip_pools": [
                ""
            ], 
            "floating_ips": null, 
            "image": "", 
            "image_exclude": "(deprecated)", 
            "key": null, 
            "key_name": "", 
            "meta": null, 
            "name": "-master", 
            "network": null, 
            "nics": [
                "net-name="
            ], 
            "region_name": null, 
            "reuse_ips": true, 
            "scheduler_hints": null, 
            "security_groups": [
                ""
            ], 
            "state": "present", 
            "terminate_volume": false, 
            "timeout": 1200, 
            "userdata": null, 
            "verify": null, 
            "volume_size": false, 
            "volumes": [], 
            "wait": true
        }
    }, 
    "msg": "Parameter 'image' or 'boot_volume' is required if state == 'present'"
}
```
Bottom line, it seems to be regression in new version, because this functionality used to work fine in previous version.

### PR DESCRIPTION
**Pull request by herlo**
**Note:** This PR will likely go through a few iterations. Specifically, the Makefile, run_tests.sh may change to accommodate testing in the containers. 

**PLEASE DO NOT APPROVE THIS PR UNTIL ALL TESTS HAVE BEEN VERIFIED MANUALLY**

Fixes #576 (now that the test cases are in place, this should be verified)
Fixes #528 

Each distro-provider combination can be tested. Certain tests can be excluded by creating a '.ignore' file. Like so: 
```
cat config/Dockerfiles/tests.d/inventory/01_template_inventory
#!/bin/bash -xe

# Verify template-based provisioning using complex template data file
# distros.exclude: none
# providers.include: libvirt aws
# providers.exclude: none

DISTROS=${1}
..snip..
```
In the above example, the `01_template_inventory` test has three commented lines. This test would run on providers libvirt and aws, all others would be excluded. No distros would be excluded. Below this comment is the code for the test

### CODE REVIEW

File: config/Dockerfiles/linchpin-tests.sh [hunk 1]
```diff
@@ -8,6 +8,12 @@ mkdir -p ${base_dir}/${distro}_logs
 
 TEST_DIR="./config/Dockerfiles/tests.d"
 
+if [ -z "${PROVIDERS}" ]; then
+    PROVIDERS="dummy openstack"
+fi
```
Comment (p3ck) [inline config/Dockerfiles/linchpin-tests.sh]: Now I'm confused.  Won't this conflict with the getTargetsToTest code in jenkins?  If anything I would think you would update that to getProvidersToTest and change the p_targets in JenkinsfileContainer to p_providers.

Comment (herlo) [reply config/Dockerfiles/linchpin-tests.sh]: This is one of the questions I had for you. I couldn't find where to update this in jenkins. I'll dig into that shortly and see if that helps. :)

File: src/org/centos/pipeline/LinchpinPipelineUtils.groovy [hunk 2]
```diff
@@ -49,6 +49,36 @@ def getTargetsToTest(targetsMap) {
...
+                    if (matchPath(file.path, e.value)) {
+                        println "${e.key} matched ${file.path}"
+                        this_match = true
+                        targets[e.key] = 1
+                    }
+                }
+                if (!this_match) {
+                    // If we get here then we have a non-target specific change
+                    // and all targets should be tested.
+                    println "Non-target file matched, will test all targets"
```
Comment (p3ck) [inline src/org/centos/pipeline/LinchpinPipelineUtils.groovy]: nit - Change target to provider

File: config/Dockerfiles/linchpin-tests.sh [hunk 3]
```diff
@@ -1,30 +1,49 @@
...
-    ./config/Dockerfiles/linchpin-test.sh $i ${distro} 2>&1 |tee ${distro}_logs/${i}.log
-    if [ $? -eq 0 ]; then
-        test_summary="$(tput setaf 2)SUCCESS$(tput sgr0)\t${testname}"
-    else
-        test_summary="$(tput setaf 1)FAILURE$(tput sgr0)\t${testname}\tlog: logs/${distro}_${i}.log"
-        result=1
-    fi
-    summary="${summary}\n${test_summary}"
+    > ${distro}_logs/${provider}.log
+    for testfile in ${TEST_DIR}/${provider}/*; do
```
Comment (p3ck) [inline config/Dockerfiles/linchpin-tests.sh]: I was hoping we could have generic tests that would be run against every provider.  If the test didn't make sense for that provider you would use the ${distro}-${provider}.ignore file.

**Comment by herlo** [comment]
@p3ck I was thinking about the per provider tests and there is a way to make a bunch of these generic. I'd like to get the basics in here and we can tweak it to have some general tests that just pass in the provider, the target(s), etc. I don't expect that to be too difficult.

File: config/Dockerfiles/linchpin-tests.sh [hunk 4]
```diff
@@ -1,31 +1,158 @@
...
+
+# this function checks the distros.exclude value within the test.
+# If none, false is returned and the disro hasn't been excluded.
+# If distros.exclude includes a distro name, it is chedked against
+# the ${distro} env var. If they match true is sent back to the caller
+# and the test will be skipped for that distro.
+function check_distro_exclude () {
+    shopt -s extglob
+    test=${1}
+    dist_exc=$(grep 'distros.exclude' ${test} | grep -v 'none' | awk -F':' '{ print $2 }')
```
Comment (p3ck) [inline config/Dockerfiles/linchpin-tests.sh]: I like the way this works.  I think it should be more strict though?  Maybe '^# distros.exclude:' ?  Same for the inlcudes below...

File: config/Dockerfiles/tests.d/openstack/05_os-vol-new [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -xe
+
+DISTRO=${1}
+PROVIDER=${2}
+
+PINFILE="PinFile.${PROVIDER}.yml"
+TARGET="os-vol-new"
+TEMPLATE_DATA="{\"distro\": \"${DISTRO}-\"}"
+TMP_PF="/tmp/PinFile.${TARGET}.out"
```
[the remark this answers is not in the document]
Comment (p3ck) [reply config/Dockerfiles/tests.d/openstack/05_os-vol-new]: you gave a thumbs up so can we change this? :)

File: config/Dockerfiles/linchpin-tests.sh (hunk 4 above)
Comment (herlo) [reply config/Dockerfiles/linchpin-tests.sh]: Ok, that sounds good. I'll update the regexp on that one. 

On the TMP_PF below, I've updated those in my latest commits, just haven't pushed them yet, so you should see that change shortly.

**Comment by herlo** [comment]
@p3ck It looks like all of the tests now pass. Check it out! 

https://jenkins-continuous-infra.apps.ci.centos.org/job/ci-linchpin/72/

I've gone through them manually, and wanted to get your take on it.

**Comment by herlo** [comment]
@p3ck, It looks like I still need to fixup the beaker and some openstack targets fail. Fixing those for the internal tests first.

**Comment by herlo** [comment]
[test]

**Comment by herlo** [comment]
[test]

**Comment by herlo** [comment]
This code now passes all of the tests within the new framework. Should be ready to merge. Requesting review from @p3ck and @samvarankashyap. Please make sure to look at the tests above and verify they are all passing as expected.

File: config/Dockerfiles/Makefile [hunk 6]
```diff
@@ -1,15 +1,15 @@
 PARENT_DIR := $(abspath $(dir $(lastword $(MAKEFILE_LIST)))/../../)
 
-TARGETS := fedora27 centos7 fedora26
+DISTROS := fedora27 centos7 fedora26
```
Comment (samvarankashyap) [inline config/Dockerfiles/Makefile]: aren't we doing fedora28 yet ?

Comment (samvarankashyap) [reply config/Dockerfiles/Makefile]: quick question . what version of centos 7.x are we using here ? is it the latest 7.5 ?

Comment (herlo) [reply config/Dockerfiles/Makefile]: not yet. That will come after this is in place. There is no container for it yet.

Comment (herlo) [reply config/Dockerfiles/Makefile]: For centos, it's whatever is in the container. Off-hand, I don't know specifically, but I think it's at least 7.4

Comment (herlo) [reply config/Dockerfiles/Makefile]: @samvarankashyap for reference, you may wish to read the [README](https://github.com/CentOS-PaaS-SIG/linchpin/blob/develop/config/README.rst) which details how the DISTRO/Dockerfile structure works.

File: config/Dockerfiles/Jenkinsfile [hunk 7]
```diff
@@ -14,17 +14,20 @@ env.FEDORA27_TAG = env.FEDORA27_TAG ?: 'stable'
...
-// if LINCHPIN_TARGETS is defined in jenkins global
-// environment this will limit the possible targets tested.
-// if it's not defined you will get all the possible targets
+// if LINCHPIN_PROVIDERS is defined in jenkins global
+// environment this will limit the possible providers tested.
+// if it's not defined you will get all the possible providers
 // listed here.
-LTARGETS = env.LINCHPIN_TARGETS ?: 'dummy duffy libvirt aws-ec2-new'
-PTARGETS = LTARGETS.tokenize()
+// LPROVIDERS = env.LINCHPIN_PROVIDERS ?: 'dummy duffy libvirt aws'
```
Comment (p3ck) [inline config/Dockerfiles/Jenkinsfile]: delete this commented dupe

File: config/Dockerfiles/JenkinsfileContainer [hunk 8]
```diff
@@ -38,19 +38,61 @@ p_targets = ['dummy':
...
-              'os-server-new':
+             'os-server-new':
               ["^linchpin/provision/roles/openstack/.*",
                "^linchpin/provision/openstack.yml"],
-              'os-sg-new':
+             'os-sg-new':
               ["^linchpin/provision/roles/openstack/.*",
                "^linchpin/provision/openstack.yml"],
-              'os-vol-new':
+             'os-vol-new':
```
Comment (p3ck) [inline config/Dockerfiles/JenkinsfileContainer]: I would just remove the old data.  it's in version control after all.

File: config/Dockerfiles/tests.d/openstack/01_os-server-new [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+TARGET="os-server-new"
+
+function clean_up {
+    set +e
+    linchpin -w . -p "${PINFILE}" -v destroy "${TARGET}"
+}
+trap clean_up EXIT SIGHUP SIGINT SIGTERM
+
+pushd docs/source/examples/workspace
+linchpin -w . -p "${PINFILE}" -v up "${TARGET}"
```
Comment (p3ck) [inline config/Dockerfiles/tests.d/openstack/01_os-server-new]: isn't this covered by general/01_general?

File: linchpin/cli/__init__.py [hunk 10]
```diff
@@ -382,7 +382,6 @@ def _execute_action(self, action, targets=(), run_id=None, tx_id=None):
 
         use_pinfile = True
         pf = None
-        pf_data = None
```
Comment (p3ck) [inline linchpin/cli/__init__.py]: this seems unrelated to this PR.  Do you need to rebase?

File: linchpin/provision/roles/libvirt/tasks/copy_image_src.yml [hunk 11]
```diff
@@ -1,7 +1,7 @@
 ---
 - name: set local_image_path
   set_fact:
-    local_image_path: "{{ libvirt_image_path }}"
+    local_image_path: "{{ libvirt_image_path | expanduser }}"
```
Comment (p3ck) [inline linchpin/provision/roles/libvirt/tasks/copy_image_src.yml]: again... this is from another PR...

File: config/Dockerfiles/tests.d/openstack/01_os-server-new (hunk 9 above)
Comment (herlo) [reply config/Dockerfiles/tests.d/openstack/01_os-server-new]: That it is, I'll remove this one.

File: linchpin/cli/__init__.py (hunk 10 above)
Comment (herlo) [reply linchpin/cli/__init__.py]: This one actually affects a tests I was working on, it needs to be in here as it was reverted a second time by #577. This fixes that again.

**Comment by herlo** [comment]
[test]

**Comment by herlo** [comment]
[test]

**Comment by herlo** [comment]
[test]

**Comment by herlo** [comment]
[test]

**Comment by herlo** [comment]
[merge]
